### PR TITLE
Split the code from AutoSchedule.cpp in two files:

### PR DESCRIPTION
**Pull request by benoitsteiner**
* the beam search part of the code itself remains in AutoSchedule.cpp
 * The generation of valide candidate to search over as well as their featurization moved to LoopNest.h

This will enable the development of new search strategies without having to duplicate all the code that now resides in LoopNest.h

### CODE REVIEW

File: apps/autoscheduler/LoopNest.h [hunk 1]
```diff
@@ -0,0 +1,2007 @@
...
+ * representation of a Halide schedule, and contains methods to
+ * generate candidates for scheduling as well as extract a
+ * featurization that can be used to cost each candidate. */
+
+#ifndef LOOP_NEST_H
+#define LOOP_NEST_H
+
+#include "FunctionDAG.h"
+#include "PerfectHashMap.h"
+#include <set>
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.h]: nit: C++ system includes should come first

File: apps/autoscheduler/LoopNest.h [hunk 2]
```diff
@@ -0,0 +1,2007 @@
...
+#include <vector>
+
+using std::set;
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.h]: It is not ok to put a 'using' of any sort into a .h file at global scope.

**Comment by benoitsteiner** [comment]
I wanted to keep the code as close the its original version. I've moved the implementation to the methods of LoopNest to a .cpp file.

**Comment by steven-johnson** [comment]
LGTM pending buildbots green

File: apps/autoscheduler/LoopNest.cpp [hunk 3]
```diff
@@ -0,0 +1,1861 @@
...
+    }
+    if (innermost) {
+        sites.get_or_create(stage).innermost = this;
+    }
+}
+
+// A helper for the working_set_at_task feature. Most features are
+// computed in the recursive pass 'compute_features' below, but
+// this one must be done in a second separate recursive pass.
+/*    void set_working_set_at_task_feature(int64_t working_set,
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.cpp]: This looks like a duplicate of the non-commented out function below; please remove

File: apps/autoscheduler/LoopNest.cpp [hunk 4]
```diff
@@ -0,0 +1,1861 @@
...
+                                const MachineParams &params,
+                                const StageMap<Sites> &sites,
+                                int64_t instances,
+                                int64_t parallelism,
+                                const LoopNest *parent,
+                                const LoopNest *grandparent,
+                                const LoopNest &root,
+                                int64_t *working_set,
+                                StageMap<ScheduleFeatures> *features) const {
+
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.cpp]: Nit: no blank line before first line of function body

File: apps/autoscheduler/LoopNest.cpp [hunk 5]
```diff
@@ -0,0 +1,1861 @@
...
+    int64_t working_set_here = 0;
+
+    int64_t loop_instances = 1, parallel_tasks = 1;
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.cpp]: Not for this PR, but perhaps a subsequent one: this is a pretty huge function body; could it be split up into smaller helper functions? (Might make it easier for other readers to reason about, and also make the compiler's job easier)

File: apps/autoscheduler/LoopNest.cpp [hunk 6]
```diff
@@ -0,0 +1,1861 @@
...
+                            if (!producer_has_been_scheduled) {
+                                // Operate optimistically and just
+                                // see if *any* dimension of the
+                                // producer would make for a good
+                                // load.
+                                int count[5] = {0, 0, 0, 0, 0};
+                                for (int i = 0; i < e->producer->dimensions; i++) {
+                                    auto stride = jac.first(i, vectorized_loop_index);
+                                    // stride is a rational. Check to see if it's a small integer.
+                                    if (stride == 0)
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.cpp]: Nit: always enclose if statements in {braces}, even single-line ones

File: apps/autoscheduler/LoopNest.cpp [hunk 7]
```diff
@@ -0,0 +1,1861 @@
...
+    f->required_to_computed(&(bound->region_required(0)), &(bound->region_computed(0)));
+
+    // Finally, figure out what loop nests will be used to compute
+    // this region.
+    for (int i = 0; i < (int)f->stages.size(); i++) {
+        f->loop_nest_for_region(i, &(bound->region_computed(0)), &(bound->loops(i, 0)));
+    }
+
+    const Bound &b = set_bounds(f, bound);
+    // b->validate();
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.cpp]: remove or justify keeping

File: apps/autoscheduler/LoopNest.cpp [hunk 8]
```diff
@@ -0,0 +1,1861 @@
...
+        }
+
+        // Uncomment when debugging the representative loop bounds selected.
+        /*
+            const auto &bounds = get_bounds(node);
+            for (size_t i = 0; i < size.size(); i++) {
+                const auto &p = bounds->loops(stage->index, i);
+                aslog(0) << " [" << p.first << ", " << p.second << "]";
+            }
+            */
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.cpp]: Nit: align the /* and */

File: apps/autoscheduler/LoopNest.cpp [hunk 9]
```diff
@@ -0,0 +1,1861 @@
...
+        // Push the Func further inwards in the loop nest
+
+        // See if it's appropriate to slide over this loop Can't
+        // slide at the root level if we intend to parallelize it.
+        // HACK
+        // bool may_slide = false;
+        bool may_slide = (params.parallelism == 1) || !is_root();
+
+        const auto &c = children[child];
+        int num_ones =
```
Comment (steven-johnson) [inline apps/autoscheduler/LoopNest.cpp]: remove or justify leaving it in (but commented out)

File: apps/support/autoscheduler.inc [hunk 10]
```diff
@@ -54,6 +54,8 @@ $(AUTOSCHED_BIN)/libauto_schedule.so: $(AUTOSCHED_SRC)/AutoSchedule.cpp \
 										$(AUTOSCHED_SRC)/Weights.cpp \
 										$(AUTOSCHED_SRC)/FunctionDAG.h \
 										$(AUTOSCHED_SRC)/FunctionDAG.cpp \
+										$(AUTOSCHED_SRC)/LoopNest.h \
+										$(AUTOSCHED_SRC)/LoopNest.cpp \
```
Comment (steven-johnson) [inline apps/support/autoscheduler.inc]: CMakeLists.txt needs this added as well

Comment (benoitsteiner) [reply apps/support/autoscheduler.inc]: Good catch, done

File: apps/autoscheduler/LoopNest.cpp (hunk 9 above)
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: I deleted it.

File: apps/autoscheduler/LoopNest.cpp (hunk 8 above)
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: done

File: apps/autoscheduler/LoopNest.cpp (hunk 7 above)
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: I added a comment about the cost of the calls to validate, as done at the beginning of the function.,

File: apps/autoscheduler/LoopNest.cpp (hunk 6 above)
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: done

File: apps/autoscheduler/LoopNest.cpp (hunk 5 above)
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: Agreed, I'll look into cleaning this up once I get to submitting further changes to this part of the code.

File: apps/autoscheduler/LoopNest.cpp (hunk 4 above)
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: done

File: apps/autoscheduler/LoopNest.cpp (hunk 3 above)
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: deleted

File: apps/autoscheduler/LoopNest.cpp [hunk 11]
```diff
@@ -0,0 +1,1861 @@
...
+namespace Halide {
+namespace Internal {
+namespace Autoscheduler {
+
+// How small should an innermost loop cluster be before you just
+// entirely unroll the thing. Sized for an architecture with 16 vector
+// registers.
+const int kUnrollLimit = 12;
+
+//template<typename T>
```
[the remark this answers is not in the document]
Comment (benoitsteiner) [reply apps/autoscheduler/LoopNest.cpp]: Deleted

**Pull request merged**